packages/eslint-plugin/src/rules/restrict-plus-operands.ts

```diff
@@ -27,166 +28,204 @@ export default util.createRule<Options, MessageIds>({
     },
     schema: [
       {
         type: 'object',
         additionalProperties: false,
         properties: {
-          checkCompoundAssignments: {
-            description: 'Whether to check compound assignments such as `+=`.',
-            type: 'boolean',
-          },
           allowAny: {
             description: 'Whether to allow `any` typed values.',
             type: 'boolean',
           },
+          allowBoolean: {
+            description: 'Whether to allow `boolean` typed values.',
+            type: 'boolean',
+          },
+          allowNullish: {
+            description:
+              'Whether to allow potentially `null` or `undefined` typed values.',
+            type: 'boolean',
+          },
+          allowNumberAndString: {
+            description:
+              'Whether to allow `bigint`/`number` typed values and `string` typed values to be added together.',
+            type: 'boolean',
+          },
+          allowRegExp: {
+            description: 'Whether to allow `regexp` typed values.',
+            type: 'boolean',
+          },
+          checkCompoundAssignments: {
+            description: 'Whether to check compound assignments such as `+=`.',
+            type: 'boolean',
+          },
         },
       },
     ],
   },
   defaultOptions: [
     {
       checkCompoundAssignments: false,
-      allowAny: false,
     },
   ],
-  create(context, [{ checkCompoundAssignments, allowAny }]) {
+  create(
+    context,
+    [
+      {
+        checkCompoundAssignments,
+        allowAny,
+        allowBoolean,
+        allowNullish,
+        allowNumberAndString,
+        allowRegExp,
+      },
+    ],
+  ) {
     const service = util.getParserServices(context);
     const typeChecker = service.program.getTypeChecker();
 
-    type BaseLiteral = 'string' | 'number' | 'bigint' | 'invalid' | 'any';
-
-    /**
-     * Helper function to get base type of node
-     */
-    function getBaseTypeOfLiteralType(type: ts.Type): BaseLiteral {
-      if (type.isNumberLiteral()) {
-        return 'number';
-      }
-      if (
-        type.isStringLiteral() ||
-        util.isTypeFlagSet(type, ts.TypeFlags.TemplateLiteral)
-      ) {
-        return 'string';
-      }
-      // is BigIntLiteral
-      if (type.flags & ts.TypeFlags.BigIntLiteral) {
-        return 'bigint';
-      }
-      if (type.isUnion()) {
-        const types = type.types.map(getBaseTypeOfLiteralType);
-
-        return types.every(value => value === types[0]) ? types[0] : 'invalid';
-      }
-
-      if (type.isIntersection()) {
-        const types = type.types.map(getBaseTypeOfLiteralType);
-
-        if (types.some(value => value === 'string')) {
-          return 'string';
-        }
-
-        if (types.some(value => value === 'number')) {
-          return 'number';
-        }
-
-        if (types.some(value => value === 'bigint')) {
-          return 'bigint';
-        }
```

Comment:
These three inner `if (types.some(...` cases weren't covered by code coverage. Most of the intersections in the previous test suite reduced to `never`. I got rid of most of them.